src/v/raft/consensus.cc

```diff
@@ -554,6 +553,13 @@ void consensus::successfull_append_entries_reply(
     idx.match_index = idx.last_dirty_log_index;
     idx.next_index = model::next_offset(idx.last_dirty_log_index);
     idx.last_successful_received_seq = idx.last_received_seq;
+    /**
+     * Update expected log end offset only if it is smaller than current value,
+     * the check is needed here as there might be pending append entries
+     * requests that were not yet replied by the follower.
+     */
+    idx.expected_log_end_offset = std::max(
+      idx.last_dirty_log_index, idx.expected_log_end_offset);
```

Comment:
I find it a bit suspicious that we are using the follower-side offset here without any term check (whereas in other places this is updated with a leader-side offset).

E.g. if both the follower and the leader have dirty_offset = 10 and term at offset 10 is different, but at offset 9 matches, and we send an append_entries with 9, the reply will be successful and expected_log_end_offset will be set to 10, but we are not really ready to send regular replicate append_entries yet.

Reply:
this is done only for successful response here, in this case follower and leader logs up to the point indicated by `append_entries_response` match perfectly